### PR TITLE
Actuator upv

### PR DESCRIPTION
**Pull request by betobetodev**

### CODE REVIEW

File: bigsea-controller/service/api/actuator/plugins/kvm_upv.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        virsh_list = "virsh list"
+        command = ("ssh root@niebla.i3m.upv.es \'ssh %s \"%s\"\'" % (host, virsh_list))
+
+        # List all the vms to get the ONE id and map with the KVM id
+        stdin, stdout, stderr = self.conn.exec_command(command)
+        vm_list = stdout.read().split("\n")
+        virsh_id = self._extract_id(vm_list, vm_id)
+
+        virsh_cap = "virsh schedinfo %s | awk 'FNR == 3 {print $3}'" % virsh_id
+        command = ("ssh root@niebla.i3m.upv.es \'ssh %s \'%s\'\'" % (host, virsh_cap))
```
Comment (armstrongmsg) [inline bigsea-controller/service/api/actuator/plugins/kvm_upv.py]: hardcoded. Maybe read "root@niebla.i3m.upv.es" from a configuration file.

**Review by tellesnobrega — COMMENTED** [review]
Looks good to me. But I will let armstrong approve it, since he knows this actuator code better than anyone

File: bigsea-controller/service/api/actuator/plugins/kvm_upv.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+    def __init__(self, conn, iops_reference, bs_reference):
+        self.conn = conn
+        self.config = ConfigParser.RawConfigParser()
+        self.config.read("controller.cfg")
+        self.one_user = self.config.get("actuator", "one_username")
+        self.one_password = self.config.get("actuator", "one_password")
+        self.one_url = self.config.get("actuator", "one_url")
+	self.iops_reference = iops_reference
+	self.bs_reference = bs_reference
```
Comment (tellesnobrega) [inline bigsea-controller/service/api/actuator/plugins/kvm_upv.py]: Indentation here is wrong

File: bigsea-controller/service/api/actuator/plugins/kvm_upv.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        command_iops_quota = (cap*self.iops_reference)/100
+        command_bs_quota = (cap*self.bs_reference)/100
+
+        command_set_io_quota = "virsh blkdeviotune %s %s --current --total_iops_sec %s --total_bytes_sec %s" % (virsh_id, 
+							block_device, command_iops_quota, command_bs_quota)
+	command = ("ssh root@niebla.i3m.upv.es \'ssh %s \'%s\'\'" % (host, command_set_io_quota))
+	
+	print "_change_io_quota: id: %s - iops: %s - bs: %s" % (virsh_id, 
+				str(command_iops_quota), str(command_bs_quota))
+
```
Comment (tellesnobrega) [inline bigsea-controller/service/api/actuator/plugins/kvm_upv.py]: Indentation here is all messed up. The whole method

**Review by tellesnobrega — COMMENTED** [review]
Please fix indentation problems, otherwise looks good

File: bigsea-controller/service/api/actuator/plugins/kvm_upv.py (hunk 2 above)
Comment (betobetodev) [reply bigsea-controller/service/api/actuator/plugins/kvm_upv.py]: Indentation here is wrong

File: bigsea-controller/service/api/actuator/plugins/kvm_upv.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+	# Get device to set cap
+	command_get_block_device = "virsh domblklist %s | awk 'FNR == 3 {print $1}'" % (virsh_id)
+	command = ("ssh root@niebla.i3m.upv.es \'ssh %s \'%s\'\'" % (host, command_get_block_device))
+	stdin, stdout, stderr = self.conn.exec_command(command)
+	block_device = stdout.read().strip()
+
+        command_iops_quota = (cap*self.iops_reference)/100
+        command_bs_quota = (cap*self.bs_reference)/100
+
+        command_set_io_quota = "virsh blkdeviotune %s %s --current --total_iops_sec %s --total_bytes_sec %s" % (virsh_id,
```
Comment (betobetodev) [inline bigsea-controller/service/api/actuator/plugins/kvm_upv.py]: For better viewing, fix the number of characters per line.

File: bigsea-controller/service/api/actuator/plugins/kvm_upv.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+        self.one_url = self.config.get("actuator", "one_url")
+	self.iops_reference = iops_reference
+	self.bs_reference = bs_reference
+
+    # TODO: validation
+    def prepare_environment(self, vm_data):
+        self.adjust_resources(vm_data)
+
+
+    # TODO: validation
```
Comment (betobetodev) [inline bigsea-controller/service/api/actuator/plugins/kvm_upv.py]: It's still TODO?

File: bigsea-controller/service/api/actuator/plugins/kvm_upv.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+        host = self._find_host(vm_id)
+
+        # List all the vms to get the ONE id and map with the KVM id
+        virsh_list = "virsh list"
+        command = ("ssh root@niebla.i3m.upv.es \'ssh %s \"%s\"\'" % (host, virsh_list))
+        stdin, stdout, stderr = self.conn.exec_command(command)
+
+        vm_list = stdout.read().split("\n")
+        virsh_id = self._extract_id(vm_list, vm_id)
+        virsh_schedinfo = (("virsh schedinfo %s | grep vcpu_quota " % (virsh_id)) + "| awk \'{print $3}\'")
```
Comment (betobetodev) [inline bigsea-controller/service/api/actuator/plugins/kvm_upv.py]: PEP8 number of characters per line.

File: bigsea-controller/service/api/actuator/plugins/kvm_upv.py (hunk 5 above)
Comment (armstrongmsg) [reply bigsea-controller/service/api/actuator/plugins/kvm_upv.py]: Yes.

File: bigsea-controller/service/api/actuator/plugins/kvm_upv.py (hunk 2 above)
Comment (armstrongmsg) [reply bigsea-controller/service/api/actuator/plugins/kvm_upv.py]: Fixed.

File: bigsea-controller/service/api/actuator/plugins/kvm_upv.py (hunk 6 above)
Comment (armstrongmsg) [reply bigsea-controller/service/api/actuator/plugins/kvm_upv.py]: OK.

File: bigsea-controller/service/api/actuator/plugins/kvm_upv.py (hunk 4 above)
Comment (armstrongmsg) [reply bigsea-controller/service/api/actuator/plugins/kvm_upv.py]: OK.

File: bigsea-controller/service/api/actuator/plugins/kvm_upv.py (hunk 3 above)
Comment (armstrongmsg) [reply bigsea-controller/service/api/actuator/plugins/kvm_upv.py]: Fixed.

**Review by iurygregory — CHANGES_REQUESTED** [review]
Many pep8 problems in the module, please fix it
http://etherpad.lsd.ufcg.edu.br/p/controller_pep8

File: bigsea-controller/service/api/actuator/plugins/kvm_upv.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+        virsh_list = "virsh list"
+        command = ("ssh root@niebla.i3m.upv.es \'ssh %s \"%s\"\'" % (host, virsh_list))
+        stdin, stdout, stderr = self.conn.exec_command(command)
+
+        vm_list = stdout.read().split("\n")
+        virsh_id = self._extract_id(vm_list, vm_id)
+        virsh_schedinfo = (("virsh schedinfo %s | grep vcpu_quota " % (virsh_id)) + 
+                                                             "| awk \'{print $3}\'")
+
+        command = ("ssh root@niebla.i3m.upv.es \'ssh %s \'%s\'\'" % (host, virsh_schedinfo))
```
Comment (igornatanael) [inline bigsea-controller/service/api/actuator/plugins/kvm_upv.py]: We could move this code to some parameter in the configuration like: compute-nodes-frontend

It is not good to use hacks like this but in this case is completely necessary, so we can minimize the ugliness of the code organizing better things like this one. :-)

File: bigsea-controller/service/api/actuator/plugins/kvm_upv.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        virsh_cap = "virsh schedinfo %s | awk 'FNR == 3 {print $3}'" % virsh_id
+        command = ("ssh root@niebla.i3m.upv.es \'ssh %s \'%s\'\'" % (host, virsh_cap))
+        stdin, stdout, stderr = self.conn.exec_command(command)
+        vm_cap = int(stdout.read().replace('\n',''))
+
+        virsh_schedinfo = (("virsh schedinfo %s" % virsh_id) 
+                         + (" --set vcpu_quota=$(( %s * 1000 ))" 
+                                % (cap)) + " > /dev/null")
+
+        command = ("ssh root@niebla.i3m.upv.es \'ssh %s \'%s\'\'" % (host, virsh_schedinfo))
```
Comment (igornatanael) [inline bigsea-controller/service/api/actuator/plugins/kvm_upv.py]: same here

**Review by igornatanael — CHANGES_REQUESTED** [review]

File: bigsea-controller/service/api/actuator/plugins/kvm_upv.py (hunk 7 above)
Comment (armstrongmsg) [reply bigsea-controller/service/api/actuator/plugins/kvm_upv.py]: done

File: bigsea-controller/service/api/actuator/plugins/kvm_upv.py (hunk 8 above)
Comment (armstrongmsg) [reply bigsea-controller/service/api/actuator/plugins/kvm_upv.py]: done

**Review by igornatanael — APPROVED** [review]

**Pull request merged**